File: go/117_populating_next_right_pointers_in_each_node_ii/main.js
function Node(val,left,right,next) {
  this.val = val;
  this.left = left;
  this.right = right;
  this.next = next;
};

// code
/**
 * // Definition for a Node.
 * function Node(val,left,right,next) {
 *    this.val = val;
 *    this.left = left;
 *    this.right = right;
 *    this.next = next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */
let connect = function(root) {
  if (null == root) return root;

  let dummyHead = new Node();
  dummyHead.next = root;
  while (dummyHead.next) {
    const nextHead = new Node();
    let tmp = nextHead;

    curr = dummyHead.next;
    while (curr) {
      if (curr.left) {
        tmp.next = curr.left;
        tmp = tmp.next;
      }
      if (curr.right) {
        tmp.next = curr.right;
        tmp = tmp.next;
      }
      curr = curr.next;
    }

    dummyHead = nextHead;
  }
  
  return root; 
};

File: go/117_populating_next_right_pointers_in_each_node_ii/main.go
package main

type Node struct {
	Val   int
	Left  *Node
	Right *Node
	Next  *Node
}

// code
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Left *Node
 *     Right *Node
 *     Next *Node
 * }
 */

func connect(root *Node) *Node {
	node := root
	for node != nil {
		iter := node

		var pre *Node
		var nextNode *Node
		for iter != nil {
			if iter.Left != nil {
				if nextNode == nil {
					nextNode = iter.Left
				}

				if pre == nil {
					pre = iter.Left
				} else {
					pre.Next = iter.Left
					pre = pre.Next
				}
			}

			if iter.Right != nil {
				if nextNode == nil {
					nextNode = iter.Right
				}

				if pre == nil {
					pre = iter.Right
				} else {
					pre.Next = iter.Right
					pre = pre.Next
				}
			}

			iter = iter.Next
		}

		node = nextNode
	}

	return root
}
